passes/opt/opt_merge.cc

```diff
@@ -275,36 +307,34 @@ struct OptMergeWorker
 				if (cell->type == ID($scopeinfo))
 					continue;
 
-				uint64_t hash = hash_cell_parameters_and_connections(cell);
-				auto r = sharemap.insert(std::make_pair(hash, cell));
-				if (!r.second) {
-					if (compare_cell_parameters_and_connections(cell, r.first->second)) {
-						if (cell->has_keep_attr()) {
-							if (r.first->second->has_keep_attr())
-								continue;
-							std::swap(r.first->second, cell);
-						}
-
+				auto [cell_in_map, inserted] = known_cells.insert(cell);
+				if (!inserted) {
+					// We've failed to insert since we already have an equivalent cell
+					Cell* other_cell = *cell_in_map;
+					if (cell->has_keep_attr()) {
+						if (other_cell->has_keep_attr())
+							continue;
+						std::swap(other_cell, cell);
```

Comment:
When this happens I think we need to update known_cells so that it holds the cell which survives

Reply:
Maybe this has been always a bug?

Reply:
Nice catch! After the loop you can add
```
for (auto cell : known_cells) {
	log("known_cells %s\n", log_id(cell));
}
```
and trip valgrind on `tests/opt/opt_merge_keep.ys`... that's no good